File: web/lib/review.ts
import { supabaseServer } from "./supabase/server";

export type ReviewStrategy =
  | "wrong-recent"
  | "frequent-miss"
  | "stale";

export const REVIEW_STRATEGIES: ReviewStrategy[] = [
  "wrong-recent",
  "frequent-miss",
  "stale",
];

export type AttemptRow = {
  question_id: string;
  correct: boolean;
  attempted_at: string;
};

async function fetchAttempts(userId: string): Promise<AttemptRow[]> {
  const sb = await supabaseServer();
  const { data, error } = await sb
    .from("attempts")
    .select("question_id, correct, attempted_at")
    .eq("user_id", userId)
    .order("attempted_at", { ascending: true });
  if (error) throw error;
  return (data ?? []) as AttemptRow[];
}

const MS_DAY = 86_400_000;

export function selectReviewCandidates(
  rows: AttemptRow[],
  strategy: ReviewStrategy,
  now: number = Date.now(),
): string[] {
  const byQ = new Map<string, AttemptRow[]>();
  for (const r of rows) {
    const arr = byQ.get(r.question_id) ?? [];
    arr.push(r);
    byQ.set(r.question_id, arr);
  }

  const ids: string[] = [];
  for (const [qid, attempts] of byQ) {
    const latest = attempts[attempts.length - 1];
    const latestMs = new Date(latest.attempted_at).getTime();

    if (strategy === "wrong-recent") {
      if (!latest.correct && now - latestMs <= 14 * MS_DAY) ids.push(qid);
    } else if (strategy === "frequent-miss") {
      const misses = attempts.filter((a) => !a.correct).length;
      if (misses >= 2) ids.push(qid);
    } else if (strategy === "stale") {
      if (now - latestMs > 7 * MS_DAY) ids.push(qid);
    }
  }
  return ids;
}

export async function getReviewCandidates(
  userId: string,
  strategy: ReviewStrategy,
): Promise<string[]> {
  const rows = await fetchAttempts(userId);
  return selectReviewCandidates(rows, strategy);
}

export async function getAllCandidateCounts(
  userId: string,
): Promise<Record<ReviewStrategy, number>> {
  const rows = await fetchAttempts(userId);
  return {
    "wrong-recent": selectReviewCandidates(rows, "wrong-recent").length,
    "frequent-miss": selectReviewCandidates(rows, "frequent-miss").length,
    stale: selectReviewCandidates(rows, "stale").length,
  };
}

File: passnote_ios/passnote_ios/Services/Review.swift
import Foundation

/// Review-mode candidate selection. Mirrors `web/lib/review.ts` so the iOS
/// and web apps surface the same questions for "review" sessions.
enum Review {
    enum Strategy: String, CaseIterable, Sendable, Identifiable {
        case wrongRecent = "wrong-recent"
        case frequentMiss = "frequent-miss"
        case stale

        var id: String { rawValue }

        var title: String {
            switch self {
            case .wrongRecent: "最近の間違い"
            case .frequentMiss: "繰り返し間違える"
            case .stale: "しばらく解いてない"
            }
        }
        var description: String {
            switch self {
            case .wrongRecent: "ここ 14 日以内に間違えた問題"
            case .frequentMiss: "通算で 2 回以上間違えた問題"
            case .stale: "7 日以上触れていない問題"
            }
        }
        var icon: String {
            switch self {
            case .wrongRecent: "exclamationmark.triangle.fill"
            case .frequentMiss: "arrow.counterclockwise"
            case .stale: "clock.badge.exclamationmark"
            }
        }
    }

    static let dayMs: TimeInterval = 86_400

    static func candidates(
        rows: [Stats.AttemptRow],
        strategy: Strategy,
        now: Date = .now,
    ) -> [String] {
        var byQ: [String: [Stats.AttemptRow]] = [:]
        for r in rows { byQ[r.questionId, default: []].append(r) }

        var ids: [String] = []
        for (qid, attempts) in byQ {
            guard let latest = attempts.last else { continue }
            let elapsed = now.timeIntervalSince(latest.attemptedAt)

            switch strategy {
            case .wrongRecent:
                if !latest.correct, elapsed <= 14 * dayMs { ids.append(qid) }
            case .frequentMiss:
                let misses = attempts.lazy.filter { !$0.correct }.count
                if misses >= 2 { ids.append(qid) }
            case .stale:
                if elapsed > 7 * dayMs { ids.append(qid) }
            }
        }
        return ids
    }

    static func counts(rows: [Stats.AttemptRow], now: Date = .now)
        -> [Strategy: Int] {
        var out: [Strategy: Int] = [:]
        for s in Strategy.allCases {
            out[s] = candidates(rows: rows, strategy: s, now: now).count
        }
        return out
    }
}
